### ISSUE TITLE
Remove Apply<TypeName> to reduce Go binary size by 70-80%

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi#6592]
As described in the analysis in https://github.com/pulumi/pulumi/issues/6591, by far the biggest impact change we could make to reduce Pulumi binary sizes is to remove the `Apply<TypeName>` methods from `*OutputState`.  

These methods contribute 80%+ of the functions total in a Pulumi program, because (a) there are 150 of them (b) they are "promoted" onto every struct that embeds `*OutputState` (c) there are two copies of each with `Context` and withtout (d) there are two more copies of each because Go promotes methods onto pointer and non-pointer receivers because these methods are called through interfaces.

These methods were added as a convenience to be a shorthand for a common pattern of usage of `ApplyT`, but their value is reasonably small.  The following are equivalent:

```go
var io pulumi.InputOutput = o.ApplyT(func(s string) (int, error) { return len(s), nil }).(pulumi.IntOutput)
var io pulumi.InputOutput = o.ApplyInt(func(s string) (int, error) { return len(s), nil })
```

The latter primarily moves where you write `Int` - in the method name instead of the type assertion (and saves a few characters because you don't have to type the `pulumi.` or `Output` parts).  

Pros of removing:
- Huge size and compilation performance reduction
- 1 fewer form of `Apply` to understand (there are 3 currently - `Apply`, `ApplyT` and `Apply<TypeName>`)
- Helper functions could be brought back later as global functions (to avoid size issues), and could even be [strongly typed](https://github.com/pulumi/pulumi/issues/6355) when brought back: `pulumi.ApplyStringToInt(o, func(s string) (int, error) { return len(s), nil })`

Cons of removing:
- Breaking change (to consider for 3.0)
- Slightly more verbose code, and perhaps more importantly, the need to remember to add a type assertion after calling `ApplyT` in most cases.

**Comment by strideynet** [comment]
This is something that leaves me quite conflicted, currently the compilation and binary sizes are very problematic, but having to remember to perform a type assertion every time you use Apply doesn't feel very idiomatic.

This leads me to suggesting that it would probably be alright to remove them, as long as  the global helper functions are introduced at the same time. I'd probably just end up personally using those, as having to cast the type is quite a faff.

**Comment by t0yv0** [comment]
Would moving and renaming these methods to having an extra dot be compromise worth considering? `pulumi.XXX.YYY(...)`. For example:

```
var io pulumi.InputOutput = o.Apply.Int(func(s string) (int, error) { return len(s), nil })
var io pulumi.InputOutput = o.ApplyInt(func(s string) (int, error) { return len(s), nil })
```

It looks a bit weird but the reason it might work out is that Go would see the .XXX as a field accessor on the struct and avoid inlining the methods, so we might keep the size/speed savings. This style might be a tad more discoverable than globals.

**Comment by elsesiy** [comment]
I'm very much in favor of this proposal as the gains far outweigh the identified downsides IMO. The UX is not impacted much by moving the type from the method signature to a type assertion. However, I do see the risk of increased panics as type assertions can easily be done "wrong".

**Comment by lukehoban** [comment]
> However, I do see the risk of increased panics as type assertions can easily be done "wrong".

One minor note - the risk of panics if the same for both of these:

```
io1 := o.ApplyT(func(s string) (int, error) { return len(s), nil }).(pulumi.StringOutput)
io2 := pulumi.InputOutput = o.ApplyString(func(s string) (int, error) { return len(s), nil })
```

Both will panic with the same error:

```
panic: interface conversion: pulumi.Output is pulumi.IntOutput, not pulumi.StringOutput
```

The latter will just report that panic from the implementation of `ApplyString` instead of from the user's code directly.

**Comment by lukehoban** [comment]
Fixed as part of the 3.0 branch with #6607.
